### ISSUE TITLE
[QUERY] DefaultAzureCredentials AuthenticationFailedException - TaskCanceled in local development

### ISSUE THREAD

**Issue by adeliab** [Azure/azure-sdk-for-net#47485]
### Library name and version

Azure.Identity 1.13.1

### Query/Question

I'm getting TaskCanceledException when using DefaultAzureCredential locally. This exception seems to happen in DefaultAzureCredential+<GetTokenFromCredentialAsync>. The exception happens locally when running the service from Visual Studio, connecting to Azurite with DefaultAzureCredential.

GetToken sometimes succeeded and sometimes failed.

These were the dependency calls related to the **failed** calls:
![Image](https://github.com/user-attachments/assets/681a08ef-8c34-4093-ad2f-129b5d1cd6f7)

These were the dependency calls related to the **successful** calls:
![Image](https://github.com/user-attachments/assets/373717e1-ff4b-42d1-ae3b-854761e5df5a)

So it doesn't succeed/ failed on specific operations. Seeing that this happens for both Blob and Table operations, I assumed the problem may be lies in Azure.Identity SDK

I've enabled Azure.Identity logs and saw that sometimes the scope requested is `https://storage.azure.com/.default` and sometimes `https://storage.azure.com.//.default`. Why does it use different scopes? I can't find `https://storage.azure.com.//.default` in azure-sdk-for-net repo
![Image](https://github.com/user-attachments/assets/9e9de9f4-350c-4487-8d53-c46252600d32)

In the above image, we see a failed calls for scope `https://storage.azure.com.//.default`. While there was a successful token for this scope at 08:35:40, expiring at 09:30:52, so why doesn't it used this token?

I saw issue [#22846](https://github.com/Azure/azure-sdk-for-net/issues/22846) but I don't think this applies in our case because we are passing the same instance of DefaultAzureCredentials to the library where this exception happens (so it's not creating a new DefaultAzureCredentials every time). 
The exception seems to be thrown in the execution of an external library that we use. Here's how they created the clients:
[TableServiceClient creation](https://github.com/petabridge/Akka.Persistence.Azure/blob/1d12d3dc1408df1e64f3f1c83490bf90eaf7f4db/src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs#L78)
[BlobServiceClient creation](https://github.com/petabridge/Akka.Persistence.Azure/blob/1d12d3dc1408df1e64f3f1c83490bf90eaf7f4db/src/Akka.Persistence.Azure/Snapshot/AzureBlobSnapshotStore.cs#L69)

The azurite setup was done with this [guide](https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=visual-studio%2Cblob-storage#azure-sdks)
The exception doesn't happen when I wrote the same code in console application.

I've also tried to increase `CredentialProcessTimeout` to 5 minutes to see if timeout is the issue, but the exception still happens

### Environment

Visual Studio 2022 version 17.11.5
.NET 8
Azurite
Azure.Storage.Blobs 12.23.0
Azure.Data.Tables 12.9.1

**Comment by christothes** [comment]
HI @adeliab - 
Can you share the full details of the exception you are seeing? Can you also share how the credential is being constructed and how its lifetime is managed?

**Comment by adeliab** [comment]
Hi @christothes ,

Here are more details on the exception:

![Image](https://github.com/user-attachments/assets/d9956102-a0aa-4b1a-bab7-480ad398073e)

### Stack trace

```
Azure.Identity.AuthenticationFailedException:
   at Azure.Identity.DefaultAzureCredential+<GetTokenFromCredentialAsync>d__13.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.DefaultAzureCredential+<GetTokenImplAsync>d__12.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at Azure.Identity.DefaultAzureCredential+<GetTokenImplAsync>d__12.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.DefaultAzureCredential+<GetTokenAsync>d__11.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+AccessTokenCache+<SetResultOnTcsFromCredentialAsync>d__9.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+AccessTokenCache+<GetAuthHeaderValueAsync>d__6.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+AccessTokenCache+TokenRequestState+<GetCurrentHeaderValue>d__19.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+AccessTokenCache+<GetAuthHeaderValueAsync>d__6.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+<AuthenticateAndAuthorizeRequestAsync>d__13.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Storage.StorageBearerTokenChallengeAuthorizationPolicy+<AuthorizeRequestInternal>d__7.MoveNext (Azure.Storage.Blobs, Version=12.23.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy+<ProcessAsync>d__12.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy+<InnerProcessAsync>d__5.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.RedirectPolicy+<ProcessAsync>d__7.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.RetryPolicy+<ProcessAsync>d__5.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.RetryPolicy+<ProcessAsync>d__5.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy+<InnerProcessAsync>d__5.MoveNext (Azure.Core, Version=1.44.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Storage.Blobs.ContainerRestClient+<GetPropertiesAsync>d__11.MoveNext (Azure.Storage.Blobs, Version=12.23.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Storage.Blobs.BlobContainerClient+<GetPropertiesInternal>d__64.MoveNext (Azure.Storage.Blobs, Version=12.23.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Storage.Blobs.BlobContainerClient+<ExistsInternal>d__61.MoveNext (Azure.Storage.Blobs, Version=12.23.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Storage.Blobs.BlobContainerClient+<ExistsAsync>d__60.MoveNext (Azure.Storage.Blobs, Version=12.23.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Akka.Persistence.Azure.Snapshot.AzureBlobSnapshotStore+<InitCloudStorage>d__13.MoveNext (Akka.Persistence.Azure, Version=1.5.31.0, Culture=neutral, PublicKeyToken=null)
Inner exception System.Threading.Tasks.TaskCanceledException handled at Azure.Identity.DefaultAzureCredential+<GetTokenFromCredentialAsync>d__13.MoveNext:
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.VisualStudioCredential+<RunProcessesAsync>d__29.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.VisualStudioCredential+<GetTokenImplAsync>d__27.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at Azure.Identity.VisualStudioCredential+<GetTokenImplAsync>d__27.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.VisualStudioCredential+<GetTokenAsync>d__25.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.ValueTask`1.get_Result (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter.GetResult (System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.DefaultAzureCredential+<GetTokenFromCredentialAsync>d__13.MoveNext (Azure.Identity, Version=1.13.1.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
```

### How the credential is constructed

The application is an ASP.NET Core API application and the credential is constructed in `Program.cs` with this code `var credentials = new DefaultAzureCredential();`
To enable logging and test the timeout setting, I changed the initialization to:
```
using var listener = AzureEventSourceListener.CreateConsoleLogger();
var credentialsOptions = new DefaultAzureCredentialOptions
{
    CredentialProcessTimeout = TimeSpan.FromSeconds(300),
    Diagnostics =
    {
        IsLoggingContentEnabled = true
    }
};
var credentials = new DefaultAzureCredential(credentialsOptions);
```

The credential works when it's deployed to cloud, but not in local development environment (tested in multiple dev machines).

### How the credential is used
This instance of credential is then used to setup multiple packages:

_Setup blob service client_
```
services.AddAzureClients(builder =>
{
    builder.UseCredential(credentials);
    builder.AddBlobServiceClient(storageSettings.EventHubCheckpointingBlobStorageEndpoint);
});
```

_Setup healthcheck_
```
builder.Services.AddHealthChecks()
    .AddAzureKeyVault(keyVaultSettings.KeyVaultUri, credentials, options => options.AddSecret($"xxx"))
```

_Setup [Akka.NET Azure persistence](https://github.com/petabridge/Akka.Persistence.Azure/)_
The exceptions seem to originate from this package - but there are also GetToken calls from this package that succeeded (as shown in my first post).

The credential is being passed on in different setting objects, for example:
```
var defaultPersistenceJournal = new AzureTableStorageJournalOptions(true)
{
    ServiceUri = storageSettings.TableStorageEndpoint,
    AzureCredential = credentials,
    TableName = "AkkaPersistenceTable",
};
var defaultPersistenceSnapshot = new AzureBlobSnapshotOptions(true)
{
    ServiceUri = storageSettings.BlobStorageEndpoint,
    AzureCredential = credentials,
    ContainerName = "akka-persistence-container",
};
```

These setting objects are then passed on to their extension methods
```
builder.Services.AddAkka(serviceName, (configurationBuilder, serviceProvider) =>
{
    configurationBuilder
       .WithAzureBlobsSnapshotStore(defaultPersistenceSnapshot)
       .WithAzureTableJournal(defaultPersistenceJournal);
}
```

### Credential lifetime
Other than initializing the credential and passing it on to these packages, we're not doing anything else with it.
I'll ask Petabridge's support if they can elaborate on their usage

**Comment by adeliab** [comment]
@christothes 

regarding the credential lifetime, this is the answer from Petabridge:
_Regarding the use of the `TokenCredential`, in the version that you're using right now, the `TableServiceClient` is being recreated (non-cached) every time an operation is performed. The `DefaultAzureCredential` passed by the user is reused throughout the whole application lifetime, it is stored inside the `_settings.AzureCredential`. This is the code where this is happening, you can share this [link](https://github.com/petabridge/Akka.Persistence.Azure/blob/375306ab5dfd740fadac71ca67eb9a0dc89c13ce/src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs#L99-L114) with the Microsoft dev team_​


> When you set the CredentialProcessTimeout, did you see longer delays before the TaskCanceled exceptions?

No it doesn't seem to have any effect.
This are the timestamps of the exception without the CredentialProcessTimeout setting:
![Image](https://github.com/user-attachments/assets/1a80ec0b-1e00-4a9d-b0d3-0752d58facf9)

And this are the timestamps with the setting set to 300 seconds:
![Image](https://github.com/user-attachments/assets/de051dc5-24c1-4992-9211-0719fe0fe1c6)


> What is strange about this behavior is that it is intermittent within a short timespan. When using a TokenCredential with one of the storage clients mentioned above, the tokens should be getting cached which would avoid calling the credential so frequently. Does this error occur only in clusters every hour or so?

Yes, that's what I was expecting too (that it won't make this many calls because of the cache). This is a local run, so I don't keep the service running all the time. The exceptions seem to happen when Akka.NET tries to initialize certain actors and it retries up to a certain point. And then Akka.NET throws its own exception
![Image](https://github.com/user-attachments/assets/434a120d-3afe-458d-b405-f5684b3fb1d2)

**Comment by christothes** [comment]
> _Regarding the use of the TokenCredential, in the version that you're using right now, the TableServiceClient is being recreated (non-cached) every time an operation is performed. 

This probably explains the reason for the issue. For `TableServiceClient`, and other SDK clients that take `TokenCredential`, they have an internal cache layer as well. Each time you create a new instance of the client you clear the cache. For credentials like `VIsualStudioCredential` and other external process based development time credentials, they need to spin up a process each time a credential is requested, except when the client using them also has a cache.

To resolve this, you'll need to somehow preserve the lifetime of the SDK clients so that they persist across invocations, or use a credential type that can persist its cache, such as `InteractiveBrowerCredential`.

**Comment by adeliab** [comment]
Hi @christothes , thanks for your reply. I will relay this to Petabridge and see what they say

**Comment by christothes** [comment]
Hi @Arkatufus - No, it is not a matter of the credential working properly. 

The issue seems to be related to the lifetime of the client SDKs using the credential. When using one of the development time credentials in the DefaultAzureCredential, such as VisualStudioCredential, a process must be started each time a token is fetched from the authority (Entra ID). However, when a client like `TableServiceClient` fetches a token from the credential, it is cached by that client for the lifetime of the token. This is normally around 1 hour. This means that the credential is only called to fetch a token, and therefore start a new process, approximately once per hour. However, when the client is not long lived and is created for each request, we never take advantage of that cache and end up calling for a token on each invocation. It appears that in this scenario, where there are multiple invocations per second, the processes start timing out. 

Ideally, each node in the system would have a singleton instance of the relevant service client such as `TableServiceClient`, that is long lived and therefore takes advantage of its token cache. 

This is typically not a problem when using one of the non-development credentials so long as the credential instance is long lived because those credentials also have a cache layer, which prevents excessive attempts to fetch new tokens. 

To summarize the guidance - a credential can be shared across any number of clients so long as they are using the same account. It is best to have singletons of any credential instance and any service client instances to minimize unnecessary token calls and reduce allocations in general.

**Comment by Arkatufus** [comment]
Hi @christothes. Thank you for the explanation, I get what the token credential is doing now, you can consider this issue as closed.
